File: eureka/android/library/src/main/java/com/lifeplus/Pojo/AServiceDef.java
package com.lifeplus.Pojo;

import com.lifeplus.Pojo.Enum.CharPropertyEnum;
import com.lifeplus.Pojo.Enum.DataTypeEnum;
import com.lifeplus.Pojo.Enum.GattCharEnum;

import java.util.HashMap;
import java.util.UUID;

public class AServiceDef {
    private final UUID _uuid;
    private final String _name;
    private final boolean _isAtomic;
    private final HashMap<UUID, ACharDef> _characteristics = new HashMap<>();

    public AServiceDef(String pUUID, String pName) {
        _uuid = UUID.fromString(pUUID);
        _name = pName;
        _isAtomic = false;
    }

    public AServiceDef(String pUUID, String pName, boolean pIsAutomic) {
        _uuid = UUID.fromString(pUUID);
        _name = pName;
        _isAtomic = pIsAutomic;
    }

    public UUID getUuid() {
        return _uuid;
    }

    public String getName() {
        return _name;
    }

    public boolean isAutomic() {
        return _isAtomic;
    }

    public void addCharacteristic(ACharDef pCharacteristic) {
        _characteristics.put(pCharacteristic.uuid, pCharacteristic);
    }

    public void addCharacteristic(String pUuid, String pName, DataTypeEnum pDataType, int pLength, CharPropertyEnum pProperty, GattCharEnum pCharEnum) {
        _characteristics.put( UUID.fromString(pUuid), new ACharDef(pUuid, pName, pDataType, pLength, pProperty, pCharEnum));
    }

    public HashMap<UUID, ACharDef> getCharacteristics() {
        return _characteristics;
    }

    public ACharDef getCharacteristic(String pUUID) {
        try {
            return _characteristics.get( UUID.fromString(pUUID));
        } catch (Throwable e) {         // Actual expected exception is IllegalArgumentException
            return null;
        }
    }

    public ACharDef getCharacteristic(UUID pUUID) {
        return _characteristics.get(pUUID);
    }
}

File: eureka/ios/nativeLib/nativeLib/Models/AServiceDef.swift
//
//  AServiceDef.swift
//  BLEDemo
//

import Foundation

public class AServiceDef {
    
    private final var _uuid: UUID
    private final var _name: String
    private final var _isAtomic: Bool
    private final var _characteristics = [UUID : ACharDef]()
    
    public init(pUUID: String, pName: String) {
        _uuid = UUID(uuidString: pUUID)!
        _name = pName
        _isAtomic = false
    }
    
    public init(pUUID: String, pName: String, pIsAutomic: Bool) {
        _uuid = UUID(uuidString: pUUID)!
        _name = pName
        _isAtomic = pIsAutomic
    }

    public func getUuid() -> UUID? {
        return _uuid
    }
    
    public func getName() -> String {
        return _name
    }
    
    public func isAutomic() -> Bool {
        return _isAtomic
    }

    public func addCharacteristic(pCharacteristic: ACharDef) {
        _characteristics[pCharacteristic.uuid!] = pCharacteristic
    }
    
    public func addCharacteristic(pUuid: String, pName: String, pDataType: DataTypeEnum, pLength: Int, pProperty: CharPropertyEnum, pCharEnum: GattCharEnum) {
        _characteristics[UUID(uuidString: pUuid)!] = ACharDef(pUuid: pUuid, pCharEnum: pCharEnum, pName: pName, pDataType: pDataType, pProperty: pProperty, pLength: pLength)
    }
    
    public func getCharacteristics() -> [UUID : ACharDef] {
        return _characteristics
    }
    
    public func getCharacteristic(pUUID: String) -> ACharDef? {
        return _characteristics.values.first(where: {
            $0.charEnum.code.lowercased() == pUUID.lowercased() || $0.charEnum.altCode.lowercased() == pUUID.lowercased()
        })
    }
}
